### ISSUE TITLE
DomainClassArtefactHandler is no longer thread safe, and can enter infinite loop

### ISSUE THREAD

**Issue by CJKirkby** [grails/grails-core#10643]
We recently upgraded to Grails 2.5.6, and yesterday we had a production server flatline at 100% CPU and needed to be terminated. Prior to killing it, we took a thread dump of the JVM process, and found 74 http-bio threads in the following state:

```
"http-bio-8080-exec-16" daemon prio=10 tid=0x00007f6c58037000 nid=0xa52 runnable [0x00007f6c50bd7000]
   java.lang.Thread.State: RUNNABLE
        at java.util.TreeMap.getEntry(TreeMap.java:353)
        at java.util.TreeMap.containsKey(TreeMap.java:227)
        at java.util.TreeSet.contains(TreeSet.java:234)
        at org.codehaus.groovy.grails.commons.DomainClassArtefactHandler.isDomainClass(DomainClassArtefactHandler.java:93)
        at org.codehaus.groovy.grails.plugins.web.api.ControllersApi.initializeCommandObject(ControllersApi.java:438)
```

After a bit of digging, I'm convinced that this is related to a change in 2.5.6 (and also appears to be in the latest release of all branches, eg. 3.1.16 & 3.2.9) that affected how DomainClassArtefactHandler caches domain vs non-domain classes. The cache storage changed from a ConcurrentHashMap into 2 TreeSets. 

The [JavaDoc for TreeSet](https://docs.oracle.com/javase/7/docs/api/java/util/TreeSet.html) is very clear that TreeSets (and the backing TreeMap) are not thread safe:

> **Note that this implementation is not synchronized.** If multiple threads access a tree set concurrently, and at least one of the threads modifies the set, it _must_ be synchronized externally.

Given that `DomainClassArtifactHandler.isDomainClass()` is accessed statically, there are bound to be concurrent modifications of the underlying TreeSets. There is a certain race condition with unsynchronized TreeSets that will create a loop in it's node graph, resulting in an infinite loop on both `add` and `contains` functionality that accesses that section of the graph.

This infinite loop behaviour can be reproduced somewhat reliably in isolation with the following unit test:

```
class TreeSetThreadSafetySpec extends Specification {

    def alphabet = (('A'..'Z')+('0'..'9')).join('')

    def "prove that concurrent modification of a TreeSet can sometimes lead to infinite loops"() {
        given: "an unsynchronized tree set"
        TreeSet<String> unsynchronizedTreeSet = [] as TreeSet
        List<Thread> threads = []

        and: "some test parameters"
        def threadCount = 10
        def stringCount = 500
        def stringLength = 3

        and: "a list of random strings using characters A-Z0-9"
        List<String> randomStrings = []
        (1..stringCount).each { randomStrings.add(generator(alphabet, stringLength)) }
        print "Finished generating ${randomStrings.size()} Strings\n"

        when: "multiple threads are started to add the strings to a set"
        ConcurrentLinkedQueue<String> addingStrings = new ConcurrentLinkedQueue<>(randomStrings.sort())
        (1..threadCount).each {
            threads << Thread.start {
                def nextString = addingStrings.poll()
                while (nextString != null) {
                    unsynchronizedTreeSet.add(nextString)
                    nextString = addingStrings.poll()
                }
                print "Add-thread ${Thread.currentThread().name} completed\n"
            }
        }

        and: "we wait a little bit to allow TreeSet to become partially "
        sleep(5000)

        and: "a single thread is started to check if TreeSet contains all strings"
        ConcurrentLinkedQueue<String> containsStrings = new ConcurrentLinkedQueue<>(randomStrings)
        threads << Thread.start {
            def nextString = containsStrings.poll()
            while (nextString != null) {
                unsynchronizedTreeSet.contains(nextString)
                nextString = containsStrings.poll()
            }
            print "Contains-thread ${Thread.currentThread().name} completed\n"
        }

        and: "we wait for all threads to complete"
        print "Waiting for all ${threads.size()} threads to complete\n"
        threads*.join()

        then: "sometimes we never reach this point"
        assert true
    }

    def generator = { String alphabet, int n ->
        new Random().with {
            (1..n).collect { alphabet[ nextInt( alphabet.length() ) ] }.join()
        }
    }

}
```

The above test _sometimes_ passes, but often there will be a NullPointerException and runs forever:

```
|Running 1 unit test... 1 of 1
--Output from prove that concurrent modification of a TreeSet can sometimes leap to infinite loops--
Finished generating 500 Strings
Error |
Exception in thread "Thread-7" 
Error |
java.lang.NullPointerException
Error |
	at java.util.TreeMap.rotateLeft(TreeMap.java:2060)
Error |
	at java.util.TreeMap.fixAfterInsertion(TreeMap.java:2127)
Error |
	at java.util.TreeMap.put(TreeMap.java:574)
Error |
	at java.util.TreeSet.add(TreeSet.java:255)
Error |
	at java_util_Set$add$1.call(Unknown Source)
Error |
	at au.com.sensis.zinc.TreeSetThreadSafetySpec$__spock_feature_0_0_closure3$_closure7.doCall(TreeSetThreadSafetySpec.groovy:32)
Error |
	at au.com.sensis.zinc.TreeSetThreadSafetySpec$__spock_feature_0_0_closure3$_closure7.doCall(TreeSetThreadSafetySpec.groovy)
Error |
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Error |
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
Error |
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Error |
	at java.lang.reflect.Method.invoke(Method.java:606)
Error |
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
Error |
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
Error |
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1218)
Error |
	at groovy.lang.ExpandoMetaClass.invokeMethod(ExpandoMetaClass.java:1125)
Error |
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1027)
Error |
	at groovy.lang.Closure.call(Closure.java:414)
Error |
	at groovy.lang.Closure.call(Closure.java:408)
Error |
	at groovy.lang.Closure.run(Closure.java:495)
Error |
	at java.lang.Thread.run(Thread.java:745)
Waiting for all 11 threads to complete
```

As recommended by the JavaDoc, this can easily be resolved by wrapping the set in a SynchronizedSortedSet:

`SortedSet s = Collections.synchronizedSortedSet(new TreeSet(...));`

I would question whether TreeSet is even appropriate as a caching mechanism in this context. Order of classes in the set does not appear to be important. This could still be implemented as 2 ConcurrentHashMap to provide the same functionality and better performance.

### Environment Information

- **Operating System**: AWS Linux 
- **Grails Version:** 2.5.6 (and likely 3.1.16 & 3.2.9)
- **JDK Version:** 1.7

**Comment by CJKirkby** [comment]
Here is my proposed solution to this issue:

https://github.com/CJKirkby/grails-core/commit/a55748be3a677ccb3cefc207b07515f442c2c901

**Comment by graemerocher** [comment]
In the end it was better to just remove the caching since reflection is no longer needed to check if a domain class is one

**Comment by CJKirkby** [comment]
Great to hear @graemerocher .

Do you plan on releasing a 2.5.x version that contains this fix?

**Comment by graemerocher** [comment]
Not unless a client specific request it no since 2.5.x is in maintenance mode. see https://objectcomputing.com/products/grails/grails-version-support-schedule/

**Comment by CJKirkby** [comment]
Thanks for the info Graeme.

Having this fixed in 2.5.x would definitely help our team, but I'm aware that we're well behind. We would love to move to the latest, but upgrading to v3 would be a significant undertaking for us right now.

If this won't be resolved, we can always drop back to v2.5.5.
